In [5]:
#Import the Pandas library
import pandas as pd
#Read the CSV data in the path provided
data = pd.read_csv(r"C:\Users\Dennis\Desktop\Employee data analysis\Employee-data-Analysis\employee_data.csv")
#Print the data to view it
data

EmpID FirstName    LastName  StartDate   ExitDate  \
0      3427     Uriah     Bridges  20-Sep-19        NaN   
1      3428     Paula       Small  11-Feb-23        NaN   
2      3429    Edward        Buck  10-Dec-18        NaN   
3      3430   Michael     Riordan  21-Jun-21        NaN   
4      3431   Jasmine       Onque  29-Jun-19        NaN   
...     ...       ...         ...        ...        ...   
2995   3422    Jakobe    Erickson  22-Jun-22  07-Aug-22   
2996   3423    Adyson  Strickland  28-Dec-20        NaN   
2997   3424   Annabel     Wilkins  09-Dec-20  04-Aug-22   
2998   3425    Kendra       Braun  28-May-19  23-Oct-21   
2999   3426     Chace        Kerr  27-Apr-22        NaN   

                        Title          Supervisor  \
0     Production Technician I        Peter Oneill   
1     Production Technician I     Renee Mccormick   
2          Area Sales Manager      Crystal Walker   
3          Area Sales Manager      Rebekah Wright   
4          Area Sales Manager           Jason Kim   
...                       ...                 ...   
2995  Production Technician I      Bethany Carter   
2996  Production Technician I     Caroline Harris   
2997  Production Technician I  Mr. James Castillo   
2998  Production Technician I       Michael Woods   
2999  Production Technician I       Javier Mccall   

                              ADEmail BusinessUnit          EmployeeStatus  \
0         uriah.bridges@bilearner.com         CCDR                  Active   
1           paula.small@bilearner.com           EW                  Active   
2           edward.buck@bilearner.com           PL                  Active   
3       michael.riordan@bilearner.com         CCDR                  Active   
4         jasmine.onque@bilearner.com          TNS                  Active   
...                               ...          ...                     ...   
2995    jakobe.erickson@bilearner.com          PYZ        Leave of Absence   
2996  adyson.strickland@bilearner.com          SVG                  Active   
2997    annabel.wilkins@bilearner.com          TNS  Voluntarily Terminated   
2998       kendra.braun@bilearner.com          WBL  Voluntarily Terminated   
2999         chace.kerr@bilearner.com          BPC                  Active   

      ...               Division         DOB State JobFunctionDescription  \
0     ...   Finance & Accounting  07-10-1969    MA             Accounting   
1     ...                 Aerial  30-08-1965    MA                  Labor   
2     ...          General - Sga  06-10-1991    MA              Assistant   
3     ...   Finance & Accounting  04-04-1998    ND                  Clerk   
4     ...          General - Con  29-08-1969    FL                Laborer   
...   ...                    ...         ...   ...                    ...   
2995  ...              Engineers  23-06-1998    MA        Model Assistant   
2996  ...              Engineers  10-06-2001    MA               Engineer   
2997  ...  Wireline Construction  27-01-1997    MA                Laborer   
2998  ...                 Aerial  28-05-1954    MA                Foreman   
2999  ...       Field Operations  11-09-1990    MA   Warehouse Technician   

     GenderCode LocationCode  RaceDesc MaritalDesc Performance Score  \
0        Female        34904     White     Widowed       Fully Meets   
1          Male         6593  Hispanic     Widowed       Fully Meets   
2          Male         2330  Hispanic     Widowed       Fully Meets   
3          Male        58782     Other      Single       Fully Meets   
4        Female        33174     Other     Married       Fully Meets   
...         ...          ...       ...         ...               ...   
2995     Female        44500     Other     Widowed       Fully Meets   
2996       Male        35444     Black     Widowed       Fully Meets   
2997       Male         6608     White      Single       Fully Meets   
2998     Female         3807  Hispanic    Divorced       Fully Meets   
2999  

In [6]:
# Convert the 'StartDate' and 'ExitDate' columns to datetime
data['StartDate'] = pd.to_datetime(data['StartDate'])
data['ExitDate'] = pd.to_datetime(data['ExitDate'])

# Calculate the tenure for each employee by subtracting 'StartDate' from 'ExitDate',
# and then dividing the result by the number of days in a year (365).
data['Tenure'] = (data['ExitDate'] - data['StartDate']).dt.days/ 365

#Calculate the average tenure of all employees
Average_tenure = data['Tenure'].mean()

#Print the average tenure in years of employees
print('The average tenure in years of employees is:',Average_tenure)

C:\Users\Dennis\AppData\Local\Temp\ipykernel_16316\416866970.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data['StartDate'] = pd.to_datetime(data['StartDate'])


The average tenure in years of employees is: 1.337318714312522


C:\Users\Dennis\AppData\Local\Temp\ipykernel_16316\416866970.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data['ExitDate'] = pd.to_datetime(data['ExitDate'])


In [8]:
#Import the visualization library
import plotly.express as px

#Calculate the counts of each unique value in the 'EmployeeStatus' column.
Status_counts = data['EmployeeStatus'].value_counts()

# Create a pie chart using Plotly Express.
#'names' parametre specifies the labels for the segment.
#'values' parametre specifies the values for each segment.
fig = px.pie(Status_counts, names=Status_counts.index ,values=Status_counts.values)

#Updatethe pie chart's traces to show both the percentage and label on the chart.
fig.update_traces(textinfo='percent+label')

#Display the pie chart
fig.show()

fig.write_image("Employee status.png")

There are 2458 employees who are currently working (active)

#**FINDING THE DISTRIBUTION OF EMPLOYEES ACROSS DIFFERENT BUSINESS UNITS**

In [10]:
data['BusinessUnit'].value_counts()

BusinessUnit
NEL     304
SVG     304
BPC     303
EW      302
PL      301
CCDR    300
PYZ     299
TNS     297
MSC     296
WBL     294
Name: count, dtype: int64

In [11]:
#Import the visualization library
import plotly.express as px

#Calculate the count of each unique value in the 'BusinessUnit' column.
units_counts = data['BusinessUnit'].value_counts()

#Create a histogam using plotly.
#'x' parametre specifies the x-value(unique business units)
#'y' parametre specifies the y-value(count of employees)
fig = px.histogram(units_counts, x=units_counts.index, y=units_counts.values)

#Update the layout. That is axis titles and the plot title.
fig.update_layout(xaxis_title = 'Business Unit', yaxis_title='Number of employees', title = 'DISTRIBUTION OF EMPLOYEES ACROSS DIFFERENT BUSINESS UNITS')

#Display the histogram
fig.show()

fig.write_image("Distribution of employees across different business units.png")

# **MOST COMMON JOB TITLE IN THE COMPANY**

In [12]:
data['Title'].value_counts()

Title
Production Technician I         1321
Production Technician II         516
Area Sales Manager               297
Production Manager               154
Software Engineer                 90
IT Support                        88
Network Engineer                  55
Sr. Network Engineer              50
Data Analyst                      47
BI Developer                      44
Administrative Assistant          33
Accountant I                      33
Sales Manager                     30
Senior BI Developer               30
Database Administrator            25
IT Manager - DB                   22
Shared Services Manager           20
Sr. Accountant                    13
IT Manager - Support              11
Director of Sales                 11
Director of Operations            11
IT Manager - Infra                11
IT Director                       11
President & CEO                   11
BI Director                       11
CIO                               11
Software Engineering Manager    

The most common job title in the company is "Production Technician I" with 1321 employees

#**DISTRIBUTION OF EMPLOYEES BASED ON THEIR EMPLOYEE TYPE**

In [13]:
data['EmployeeType'].value_counts()

EmployeeType
Full-Time    1038
Contract     1008
Part-Time     954
Name: count, dtype: int64

In [14]:
#Importing the visualizatiion library
import plotly.express as px

#Finding the counts in the 'EmployeeType' column.
Type_counts = data['EmployeeType'].value_counts()

#Create a pie chart.
#'name' parametre specifies the labels
#'Values' paremetre secifies the counts under each label
fig = px.pie(Type_counts, names=Type_counts.index, values=Type_counts.values)

#Update the chart to show both percentage and labels
fig.update_traces(textinfo='percent+label')

#Display the pie chart
fig.show()

fig.write_image("bar_chart.png")

488.12133072407045

#**DURATION OF EMPLOYMENT IN DAYS**

In [16]:
#Convert the 'StartDate' and 'ExitDate' columns to datetime.
data['StartDate'] = pd.to_datetime(data['StartDate'])
data['ExitDate'] = pd.to_datetime(data['ExitDate'])

#Calculate the duration of employment for each employee in days by subtracting 'StartDate' from 'ExitDate'
data['Employee Duration'] = (data['ExitDate'] - data['StartDate']).dt.days

#Calculate the average duration of employment for all employees.
average_duration = data['Employee Duration'].mean()

average_duration

488.12133072407045

Average days of employees are 488 days

#**NUMBER OF EMPLOYEES WHO HAVE EXITED THE COMPANY IN EACH YEAR**

In [17]:
#Convert the 'ExitDate' column to datetime.
data['ExitDate'] = pd.to_datetime(data['ExitDate'])

#Extract the year from the 'ExitDate' column and store it in the column 'ExitYear'
data['ExitYear'] = data['ExitDate'].dt.year

#Calculate the count of employee exits for each year and sort the results by year.
exit_count_by_year = data['ExitYear'].value_counts().sort_index()

exit_count_by_year

ExitYear
2018.0      4
2019.0     62
2020.0    133
2021.0    278
2022.0    460
2023.0    596
Name: count, dtype: int64

The above output shows number of employees that exited in respective years

In [18]:
#Importing the visualization library
import plotly.express as px

#Convert the 'ExitDate' column in datetime.
data['ExitDate'] = pd.to_datetime(data['ExitDate'])

#Create a histogram that shows distribution of employees according to the exit years
fig = px.histogram(data['ExitDate'])

#Update the histogram by adding title in the axes and the plot title
fig.update_layout(xaxis_title = 'Dates', yaxis_title = 'Number of employees', title='DISTRIBUTION OF EXIT DATES ON EMPLOYEES')

#Display the graph
fig.show()

fig.write_image('Distributionof exit dates on employees.png')

The trend of exit of employees increases as the year increases

#**SUPERVISOR WITH THE MOST EMPLOYEES UNDER THEM** 

In [20]:
data['Supervisor'].value_counts()

Supervisor
John Smith          3
Robert Sullivan     2
Kenneth Perry       2
Desiree Martinez    2
Susan Smith         2
                   ..
Barry Payne         1
Paula Ross          1
Lisa Spencer        1
Joseph Smith        1
Javier Mccall       1
Name: count, Length: 2952, dtype: int64

John smith is the supervisor with the most employees under him.

#**DISTRIBUTION OF EMPLOYEES BASED ON GENDER.**

In [21]:
data['GenderCode'].value_counts()

GenderCode
Female    1682
Male      1318
Name: count, dtype: int64

In [22]:
#Import the visualization library
import plotly.express as px

#Find the count of each value in the 'GenderCode' column.
gender_count = data['GenderCode'].value_counts()

#Create a piechart.
#'names' parameter specifies the labels for the segments.
#'Values' paremeter specifies the values for each segment.
#'Title' paremeter sets the title chart.
fig = px.pie(gender_count, names=gender_count.index, values=gender_count.values, title='DISTRIBUTION OF EMPLOYEES BASED ON GENDER')

#Update the pie chart's traces to show both the percentage and label on the chart
fig.update_traces(textinfo='percent+label')

#Display the pie chart
fig.show()


fig.write_image("Distribution on employees bsed on Gender.png")

Most of the employees in the company are women

**#DISTRIBUTION BASED ON THEIR RACE#**

In [25]:
data['RaceDesc'].value_counts()

RaceDesc
Asian       629
Black       618
White       599
Other       582
Hispanic    572
Name: count, dtype: int64

In [26]:
#Import the visualization library
import plotly.express as px

#Find the counts on each value in the 'RaceDesc' column.
race_count = data['RaceDesc'].value_counts()

#Create the pie chart.
#'names' parametre specifies the values for each segment.
#'title' parametre specifies the values for the chart.
fig = px.pie(race_count, names=race_count.index, values=race_count.values,labels=race_count.index, title='DISTRIBUTION OF EMLOYEES BASED ON RACE')

#Update the piechart's tarces to show both percentage and label on the chart.
fig.update_traces(textinfo='percent+label')

#Display the chart
fig.show()

fig.write_image("Distribution of employeesbased on race.png")

Most of the employees are Asian

#**DISTRIBUTION OF EMPLOYEES BASEDON THEIR MARITAL STATUS.**

In [27]:
data['MaritalDesc'].value_counts()

MaritalDesc
Single      772
Married     764
Divorced    741
Widowed     723
Name: count, dtype: int64

In [28]:
#Import the visualization library
import plotly.express as px

#Calculate the counts of each unique value in the 'MaritalDesc' column.
marital_count = data['MaritalDesc'].value_counts()

# Create a pie chart.
# 'names' parameter specifies the labels for the segments.
# 'values' parameter specifies the values for each segment.
# 'title' parameter sets the title for the chart.
fig = px.pie(marital_count, names=marital_count.index, values=marital_count.values, title='DISTRIBUTION OF EMPLOYEES BASED ON THEIR MARITAL STATUS')

# Update the pie chart's traces to show both the percentage and label on the chart.
fig.update_traces(textinfo='percent+label')

#Display the pie chart.
fig.show()

fig.write_image("Distribution of employees based on their marital status.png")

Most of the employees single in the company

#**DISTRIBUTION OF EMPLOYEE PERFOMANCE BASED ON GENDER AND RACE**

In [29]:
#Import the visualization library
import plotly.express as px

#Create a bar chart for employee perfomance score based on Race
#'x' paraemtre specifies the cartegorical variable (RaceDesc) on the x axis.
#'y' parametre specifies the categorical variable (Perfomance Score) on the y axis.
fig_race = px.bar(data, x='RaceDesc', y='Performance Score' , title='DISTRIBUTION OF EMPLOYEES PERFOMANCE SCORE BASED ON THEIR RACE')

#Create a bar chart for employee perfomance score based on Gender
#'x' paraemtre specifies the cartegorical variable (RaceDesc) on the x axis.
#'y' parametre specifies the categorical variable (Perfomance Score) on the y axis.
fig_Gender = px.bar(data, x='GenderCode', y='Performance Score', title='DISTRIBUTION OF EMPLOYEES PERFOMANCE BASED ON THEIR GENDER')

#Display both bar graphs
fig_Gender.show()
fig_race.show()

fig_Gender.write_image("Distribution of employees based on Gender.png")
fig_race.write_image("Distribution of eployee based on their Race.png")

In the graph showing distribution of employee perfomance based on their Gender, Females show a greater average perfomance than Males.

In the graph showing distribution of employee perfomance based on ther races, Blacks show a greater average perfomance more than all the other races.

In [30]:

# Group the data by gender and performance score, and calculate the count
performance_by_gender = data.groupby(['GenderCode', 'Performance Score']).size().reset_index(name='Count')

# Print the results
print(performance_by_gender)


  GenderCode  Performance Score  Count
0     Female            Exceeds    243
1     Female        Fully Meets   1278
2     Female  Needs Improvement    104
3     Female                PIP     57
4       Male            Exceeds    126
5       Male        Fully Meets   1083
6       Male  Needs Improvement     73
7       Male                PIP     36


The above output indicates that more females exceeed the perfomance score more than male, more females fully meet the perfomance score more than male, More females need improvement more than males. More female pip than males.

In [31]:
# Group the data by gender and performance score, and calculate the count
perfomance_by_race = data.groupby(['RaceDesc','Performance Score']).size().reset_index()

# Print the results
perfomance_by_race

RaceDesc  Performance Score    0
0      Asian            Exceeds   80
1      Asian        Fully Meets  499
2      Asian  Needs Improvement   36
3      Asian                PIP   14
4      Black            Exceeds   58
5      Black        Fully Meets  496
6      Black  Needs Improvement   42
7      Black                PIP   22
8   Hispanic            Exceeds   75
9   Hispanic        Fully Meets  447
10  Hispanic  Needs Improvement   27
11  Hispanic                PIP   23
12     Other            Exceeds   77
13     Other        Fully Meets  452
14     Other  Needs Improvement   34
15     Other                PIP   19
16     White            Exceeds   79
17     White        Fully Meets  467
18     White  Needs Improvement   38
19     White                PIP   15

In [32]:
import pandas as pd
import datetime
# Convert the 'DOB' column to datetime format.
data['DOB'] = pd.to_datetime(data['DOB'])

# Calculate the age for each employee by subtracting their birth year from the current year.
data['Age'] = (datetime.datetime.now().year - data['DOB'].dt.year)

# Group the data by 'DepartmentType' and calculate the average age for each group.
Average_age_by_department = data.groupby('DepartmentType')['Age'].mean()
Average_age_by_department

ValueError: time data "30-08-1965" doesn't match format "%m-%d-%Y", at position 1. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

In [33]:
#Import the visualization library
import plotly.express as px

# Create a box plot using Plotly Express.
# 'x' parameter specifies the categorical variable (EmployeeClassificationType) on the x-axis.
# 'y' parameter specifies the numerical variable (PayZone) on the y-axis.
# 'title' parameter sets the title for the chart.
fig = px.box(data, x='EmployeeClassificationType', y='PayZone', title='Pay Variation Across Employee Classification Types')

#Update the axes title in the graph
fig.update_layout(xaxis_title='Employee Classification Type', yaxis_title='Pay')

#Display the graph
fig.show()




In [34]:
import pandas as pd
import plotly.express as px

# Assuming your dataset is in a CSV file named "employee_data.csv"

# Create an interactive bar plot using Plotly
fig = px.histogram(data, x='State', title='Distribution of Employees Across Different States')
fig.update_xaxes(title='State')
fig.update_yaxes(title='Number of Employees')
fig.update_layout(xaxis_tickangle=-45)
fig.show()


fig.write_image("Distribution of employees across different States.png")

Most employees are from the state of MA